src/Microsoft.AspNetCore.Server.Kestrel/KestrelServer.cs

```diff
@@ -108,8 +114,45 @@ public void Start<TContext>(IHttpApplication<TContext> application)
                     else
                     {
                         atLeastOneListener = true;
-                        _disposables.Push(engine.CreateServer(
-                            parsedAddress));
+
+                        if (parsedAddress.Host.Equals("localhost", StringComparison.OrdinalIgnoreCase))
+                        {
+                            var exceptions = new List<Exception>();
+                            var originalHost = parsedAddress.Host;
+
+                            var loopbackAddresses = new[]
+                            {
+                                new { Address = IPAddress.Loopback.ToString(), Family = "IPv4" },
+                                new { Address = IPAddress.IPv6Loopback.ToString(), Family = "IPv6" }
+                            };
+
+                            foreach (var loopbackAddress in loopbackAddresses)
+                            {
+                                try
+                                {
+                                    parsedAddress.Host = loopbackAddress.Address;
+                                    _disposables.Push(engine.CreateServer(
+                                        parsedAddress));
+                                }
+                                catch (UvException ex) when (ex.StatusCode != Constants.EADDRINUSE)
+                                {
+                                    _logger.LogWarning(0, ex, $"Unable to bind {address} to loopback {loopbackAddress.Family} address.");
```

Comment:
`address` would be `http://localhost:someport/something`. The idea is to inform that we failed to bind to a given address when the user asked to bind to localhost.